### PR TITLE
chore: Implement strict mode

### LINKED ISSUE #213: Implement strict mode in `std`
Deno runs in non-strict mode, in the Deno core though, we enforce strict mode.  I think we should enforce that for `std`.  I am seeing a lot of code in here that is not compliant with strict mode, which allows a lot of unsafe assignments around `null` and `undefined` as well as quite a few other violations which make the code unsound.

We also have quite a few linting violations too, which we catch in core, which are not being caught in our code review, like unused variables, `let`s where no reassignment occurs, etc...

My opinion is we need to find a way to enforce these as part of the CI process before it gets too far out of hand.

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes #213

### CODE REVIEW

File: log/levels.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
-export const LogLevel = {
+export const LogLevel : {[key: string]: number } = {
```
Comment (zekth) [inline log/levels.ts]: is `Record<string,number>` equivalent?

**Comment by bartlomieju** [comment]
This task is more complicated than I originally thought; passing `tsconfig.json` via `--config` to Deno doesn't take care about `files`, `include` and `exclude` options (as it shouldn't) so we have 2 options for this PR:
- land several smaller PRs cleaning codebase gradually, PRs have to be tested locally one by one and after all landed then enable strict mode in CI
- clean everything in one PR (that is almost each file in repo has to be modified) with CI enforcing strict mode from beginning

What are your opinions?
CC @ry @kitsonk

**Comment by zekth** [comment]
Maybe add problematic files in exclude for the check and fix it in further PR? Otherwise it will be a bit too much work in the same PR i think.

**Comment by bartlomieju** [comment]
> Deno doesn't take care about `files`, `include` and `exclude` options

@zekth as stated above, Deno's TS compiler does not respect either of options

**Comment by zekth** [comment]
> > Deno doesn't take care about `files`, `include` and `exclude` options
> 
> @zekth as stated above, Deno's TS compiler does not respect either of options

I was meaning having strict_test.ts using strict.tsconfig.json for example. But we dont have to let it stay here for long.

**Comment by bartlomieju** [comment]
Got it to green, I think we can use `tsconfig.json` that is already in repo (though it will produce some warnings about ignored options). 

Let me know what you think, @ry, @kitsonk

**Comment by kitsonk** [comment]
You could consider putting a minimal tsconfig in tests and extend it in the root to avoid the ignored options. But otherwise, good work.

**Comment by bartlomieju** [comment]
@kitsonk I did as you suggested. 

It's worth noting that there is still a lot of places with `any` types ignored by eslint, but I'm sure we can tackle that gradually in separate PRs.

Kudos to @zekth for a lot of help

**Comment by bartlomieju** [comment]
Let's wait for #444 before landing this PR

**Comment by ry** [comment]
LGTM but holding for #444

**Comment by bartlomieju** [comment]
@ry I updated this PR after #444, PTAL

File: azure-pipelines.yml [hunk 2]
```diff
@@ -37,4 +37,4 @@ jobs:
       - bash: echo "##vso[task.prependpath]C:\Users\VssAdministrator\.deno\\bin"
       - bash: npx eslint **/*.ts --max-warnings=0
       - bash: deno.exe run --allow-run --allow-write --allow-read format.ts --check
-      - bash: deno.exe run --allow-run --allow-net --allow-write --allow-read test.ts
+      - bash: deno.exe run --allow-run --allow-net --allow-write --allow-read --config=tsconfig.test.json test.ts
```
Comment (ry) [inline azure-pipelines.yml]: Why is the --config flag being added here?

File: azure-pipelines.yml [hunk 3]
```diff
@@ -26,7 +26,7 @@ jobs:
       - script: echo '##vso[task.prependpath]$(HOME)/.deno/bin/'
       - script: eslint **/*.ts --max-warnings=0
       - script: deno run --allow-run --allow-write --allow-read format.ts --check
-      - script: deno run --allow-run --allow-net --allow-write --allow-read test.ts
+      - script: deno run --allow-run --allow-net --allow-write --allow-read --config=tsconfig.test.json test.ts
```
Comment (ry) [inline azure-pipelines.yml]: We need to de-dup these commands at some point...

Probably we should try to copy [how Tokio uses Azure Pipelines ](https://github.com/tokio-rs/tokio/tree/84d5a7f5a0660dbe1e2be0e831e715f13cb59a2b/ci)

File: azure-pipelines.yml (hunk 2 above)
Comment (bartlomieju) [reply azure-pipelines.yml]: Well I tried using strict mode in linter but it was not making any difference. So I explicitly tell Deno's TS compiler to use strict mode.

File: azure-pipelines.yml (hunk 3 above)
Comment (bartlomieju) [reply azure-pipelines.yml]: Definitely, I'll do that in followup PR

**Pull request merged**